pkg/logql/rangemapper.go

```diff
@@ -0,0 +1,313 @@
+	var vectorAggrPushdown *syntax.VectorAggregationExpr
+	if expr.Grouping != nil && expr.Grouping.Groups != nil && expr.Operation != syntax.OpTypeCount {
+		vectorAggrPushdown = expr
+	}
+
+	// Split the vector aggregation's inner expression
+	lhsMapped, err := m.Map(expr.Left, vectorAggrPushdown)
+	if err != nil {
+		return nil, err
+	}
+
+	return &syntax.VectorAggregationExpr{
+		Left:      lhsMapped,
+		Grouping:  expr.Grouping,
+		Params:    expr.Params,
+		Operation: expr.Operation,
+	}, nil
+}
+
+// mapRangeAggregationExpr maps expr into a new SampleExpr with multiple downstream subqueries split by range interval
+// Optimization: in order to reduce the returned stream from the inner downstream functions, in case a range aggregation
+// expression is aggregated by a vector aggregation expression with a label grouping, the downstream expression can be
+// exactly the same as the initial query concatenated by a `sum` operation. If this is the case, overrideDownstream
+// contains the initial query which will be the downstream expression with a split range interval.
+// Example: `sum by (a) (bytes_over_time)`
+// Is mapped to `sum by (a) (sum without downstream<sum by (a) (bytes_over_time)>++downstream<sum by (a) (bytes_over_time)>++...)`
+func (m RangeVectorMapper) mapRangeAggregationExpr(expr *syntax.RangeAggregationExpr, vectorAggrPushdown *syntax.VectorAggregationExpr) syntax.SampleExpr {
+	rangeInterval := getRangeInterval(expr)
+
+	// in case the interval is smaller than the configured split interval,
+	// don't split it.
+	if rangeInterval <= m.splitByInterval {
+		return expr
+	}
+
+	if expr.Grouping == nil && hasLabelExtractionStage(expr) {
+		return expr
+	}
+	switch expr.Operation {
+	case syntax.OpRangeTypeBytes, syntax.OpRangeTypeCount, syntax.OpRangeTypeSum:
+		return m.vectorAggrWithRangeDownstreams(expr, vectorAggrPushdown, syntax.OpTypeSum, rangeInterval)
+	case syntax.OpRangeTypeMax:
+		return m.vectorAggrWithRangeDownstreams(expr, vectorAggrPushdown, syntax.OpTypeMax, rangeInterval)
+	case syntax.OpRangeTypeMin:
+		return m.vectorAggrWithRangeDownstreams(expr, vectorAggrPushdown, syntax.OpTypeMin, rangeInterval)
+	case syntax.OpRangeTypeRate:
+		return m.sumOverFullRange(expr, vectorAggrPushdown, syntax.OpRangeTypeCount, rangeInterval)
+	case syntax.OpRangeTypeBytesRate:
+		return m.sumOverFullRange(expr, vectorAggrPushdown, syntax.OpRangeTypeBytes, rangeInterval)
+	default:
+		// this should not be reachable. If an operation is splittable it should
+		// have an optimization listed
+		level.Warn(util_log.Logger).Log(
+			"msg", "unexpected range aggregation expression which appears shardable, ignoring",
+			"operation", expr.Operation,
+		)
+		return expr
+	}
+}
+
+func isSplittableByRange(expr syntax.SampleExpr) bool {
+	switch e := expr.(type) {
+	case *syntax.VectorAggregationExpr:
+		_, ok := SplittableVectorOp[e.Operation]
+		return ok && isSplittableByRange(e.Left)
+	case *syntax.BinOpExpr:
+		return true
```

Comment:
Good point. Atm it works, because for each leave we call Map again and check if anything further down is splittable.
Checking also leaves recursively here will allow to call `isSplittableByRange` once, even before `Map()`.

Reply:
I'm mostly worried that the function says one thing (tests if the Expr is splittable) but doesn't always do it correctly. This can work _now_, but it seems easily susceptible to accidental refactoring which breaks.

Reply:
Yes, makes sense. You are totally right. Implemented fully recursive and moved it before the `Map()` call.